dsa.go

```diff
@@ -225,22 +233,18 @@ func newDSA1(params DSAParameters, x, y BigInt) (C.GO_EVP_PKEY_PTR, error) {
 		if ret != 1 {
 			C.go_openssl_BN_free(pub)
 			C.go_openssl_BN_clear_free(priv)
-			C.go_openssl_DSA_free(dsa)
 			return nil, newOpenSSLError("DSA_set0_key failed")
 		}
 	} else {
 		if C.go_openssl_DSA_generate_key(dsa) != 1 {
-			C.go_openssl_DSA_free(dsa)
 			return nil, newOpenSSLError("DSA_generate_key failed")
 		}
 	}
-	pkey := C.go_openssl_EVP_PKEY_new()
+	pkey = C.go_openssl_EVP_PKEY_new()
 	if pkey == nil {
-		C.go_openssl_DSA_free(dsa)
 		return nil, newOpenSSLError("EVP_PKEY_new failed")
 	}
 	if C.go_openssl_EVP_PKEY_assign(pkey, C.GO_EVP_PKEY_DSA, unsafe.Pointer(dsa)) != 1 {
-		C.go_openssl_DSA_free(dsa)
```

Comment:
Assuming the existing code was right and calling `C.go_openssl_DSA_free(dsa)` is needed in this error case, this seems wrong.

My understanding is that `EVP_PKEY_assign` is what actually transfers ownership, and when we hit this `!= 1` the ownership isn't transferred, and that's why `C.go_openssl_DSA_free(dsa)` is needed. But at this point, `pkey` isn't `nil`, so the deferred cleanup won't call `DSA_free`.

(This is something I liked about the explicit `dsa = nil // Ownership has been transferred to pkey.`--it's very clear what's intended and doesn't tie state together.)

Reply:
Oop, I missed how the change to use a named return affects this. Never mind!